Update with structural types (#15)
keizer619/release-notes: ballerina-release-notes-0.990.0.md

# Highlights

> - We’ve introduced changes to the Ballerina type system to encourage data-oriented thinking, to provide the foundation for safe concurrency by introducing immutable values, and to make structural types and mutability work together smoothly. 

Reviewer: ```suggestion
- We have introduced changes to the Ballerina type system to encourage data-oriented thinking, to provide the foundation for safe concurrency by introducing immutable values, and to make structural types and mutability work together smoothly. 
```

- We’ve introduced changes to error handling in Ballerina to maintain a consistent policy on normal and abnormal errors throughout the language. 

> - Error handling related changes affects the semantics of concurrency constructs: error types are statically propagated to interacting workers. Workers can now have a statically-typed return value which may be an error. 

Reviewer: ```suggestion
- Error handling related changes affects the semantics of concurrency constructs: error types are statically propagated to interacting workers. Workers can now have a statically-typed return value, which may be an error. 
```

- Cloning and immutability of structured values contribute to message send/receive safety across workers. 

> - Endpoints and services are the key abstractions in Ballerina that bring network programming to a higher level abstraction when compared to traditional languages. We’ve introduced significant changes to endpoints and services syntax and semantics. 

Reviewer: ```suggestion
- Endpoints and services are the key abstractions in Ballerina that bring network programming to a higher level of abstraction when compared to traditional languages. We have introduced significant changes to endpoints and services syntax and semantics. 
```

Ballerina has a structural type system in which compatibility of types and values is determined based on their structure; this is particularly useful in integration scenarios that combine multiple, independently-designed systems. The following changes to the Ballerina type system encourage data-oriented thinking, provides the foundation for safe concurrency by introducing immutable values, makes structural types, and mutability work together smoothly.  

### Typing of mutable structure values

> Ballerina programs use types to categorize values both at compile-time and runtime. Types deal with an abstraction of values, which does not consider storage location or mutability. This abstraction is called a shape.  A type denotes a set of shapes. A type S is a subtype of type T if the set of shapes denoted by S is a subset of the set of shapes denoted by T. Every value has a corresponding shape. Since shapes do not deal with storage location, they have no concept of identity; shapes therefore represent trees rather graphs. Shapes differ from values only for reference values. For simple values, there is no difference between a shape and a value; the shape of a simple value is just that value. A shape is specific to a basic type: if two values have different basic types, then they have different shapes.

Reviewer: ```suggestion
Ballerina programs use types to categorize values both at compile-time and runtime. Types deal with an abstraction of values, which does not consider storage location or mutability. This abstraction is called a shape. A type denotes a set of shapes. A type `S` is a subtype of type `T` if the set of shapes denoted by `S` is a subset of the set of shapes denoted by `T`. Every value has a corresponding shape. Since shapes do not deal with storage location, they have no concept of identity; shapes therefore represent trees rather graphs. Shapes differ from values only for reference values. For simple values, there is no difference between a shape and a value; the shape of a simple value is just that value. A shape is specific to a basic type: if two values have different basic types, they have different shapes.
```

```

### Type test expression
The type test can be used to determine whether a value is of a particular type. 
```ballerina
int|string x = getValue();
if x is string {
   // x is now string after implicit narrowing
   return;
}
// x is now int
```

### Type assertions
Type assertions can be used to assert the inherent type of a value. Assertion fails and results in a panic, if the asserted type is not exactly the storage type.
```ballerina

Employee e = { ... };
Person p = e;

Employee e2 = <Employee> p; // successful since storage type is Employee
Person p2 = <Person> p; // panics since storage type is Employee
```

> ### match statement 

Reviewer: ```suggestion
### Match statement 
```

}
	var x => io:println(“Matched to Default Pattern”); //all values will match
}
```

### Dynamically narrowing type

> We talked about two types of relationships between a value and a type in this section: looks like and belongs to. With the introduction of these two relationships, one might want to transform a value that looks like a type into a value that belongs to that type. We’ve introduced two built-in methods to do that.

Reviewer: Can we reword this? The way this is worded here makes me think that "looks like and belongs to" is the name of a section in the release notes, and that's not the case.

> - T.stamp(v): mutate the value v itself and set its inherent type to T.

Reviewer: ```suggestion
- `T.stamp(v)`: mutate the value `v` itself and set its inherent type to `T`.
```

> - T.convert(v): Make copy of the value v and  set its inherent type to T.

Reviewer: ```suggestion
- `T.convert(v)`: Make copy of the value `v` and  set its inherent type to `T`.
```

```

## Variable Initialization
### Compile time constants

> A compile-time constant is a typed identifier whose value is computed at compile time. Compile time reference values are frozen values 

Reviewer: ```suggestion
A compile-time constant is a typed identifier whose value is computed at compile time. Compile time reference values are frozen values.
```

```ballerina
const i = 10;
```

### Final variables

> The final modified can be applied to a variable declaration to mean that the variable cannot be modified after it has been initialized

Reviewer: Shall we say "Modified final variable" or something similar? The "final modified" has incomplete meaning for a newer user.

# Experimental Language Features 

> You can unlock experimental language features with the `--experimental` flag

Reviewer: ```suggestion
You can unlock experimental language features with the `--experimental` flag.
```
